File: app/src/lib/game/computer/modules/setup-plays.ts
import { isActionCard, isSuitCard, isWhotCard } from '../../guards.js';
import { getValidMoves } from '../../moves.js';
import type { TopCard } from '../../types.js';
import type { Candidate, ModuleContext, ScoringModule } from '../types.js';

// Prefer plays that leave us with more options on our next turn.
// Simulate the resulting top card, then count how many cards in our
// remaining hand would be playable against it.
export const setupPlays: ScoringModule = (candidate: Candidate, ctx: ModuleContext): number => {
	if (candidate.kind === 'draw') return 0;

	const player = ctx.state.players[ctx.playerIndex];
	if (player === undefined) return 0;

	// Simulate the top card that results from this play
	const simulatedTop: TopCard =
		candidate.kind === 'play-suit'
			? candidate.card
			: { kind: 'whot', value: 20, calledShape: candidate.calledShape };

	// Remove the played card from our hand to see what's left
	const remainingHand =
		candidate.kind === 'play-suit'
			? player.hand.filter(
					(c) =>
						!(
							isSuitCard(c) &&
							c.shape === candidate.card.shape &&
							c.value === candidate.card.value
						)
				)
			: player.hand.filter((c) => !isWhotCard(c));

	// Playing the last card — always the right move
	if (remainingHand.length === 0) return 50;

	// Count how many of our remaining cards could play against the simulated top
	const followUpMoves = getValidMoves(remainingHand, simulatedTop, null, ctx.state.mode);

	// In N-player games the top card will change N-1 times before we play again,
	// so setup quality is less predictable. Scale scores down accordingly.
	const scale = 1 / Math.max(1, ctx.state.players.length - 1);

	if (followUpMoves.length === 0) {
		// Action cards disrupt the opponent regardless of our follow-up position;
		// don't penalise them for landing on an awkward shape.
		if (candidate.kind === 'play-suit' && isActionCard(candidate.card)) return 0;
		// Non-action play that leaves us with nothing to follow — strong deterrent
		return -20 * scale;
	}

	// Cap at 6 so large hands don't overwhelm action-awareness in the score hierarchy
	return Math.min(6, followUpMoves.length * 1.5) * scale;
};

File: server/src/game/ai/modules/setup_plays.rs
use crate::game::{
    ai::types::{Candidate, ModuleContext},
    engine::is_action_card,
    moves::valid_moves,
    types::{Card, TopCard},
};

/// Prefer plays that leave us with more follow-up options next turn.
pub fn setup_plays(candidate: &Candidate, ctx: &ModuleContext<'_>) -> f64 {
    if matches!(candidate, Candidate::Draw) {
        return 0.0;
    }

    let Some(seat) = ctx.state.seats.get(ctx.seat_index) else {
        return 0.0;
    };

    let simulated_top: TopCard = match candidate {
        Candidate::PlaySuit { shape, value } => TopCard::Suit {
            shape: *shape,
            value: *value,
        },
        // Groups are scored via their representative single in params.rs, so this
        // module never actually receives one; treat it as its top card anyway.
        Candidate::PlayGroup { value, top_shape, .. } => TopCard::Suit {
            shape: *top_shape,
            value: *value,
        },
        Candidate::PlayWhot { called_shape } => TopCard::Whot {
            called_shape: *called_shape,
        },
        Candidate::Draw => unreachable!(),
    };

    // Remove exactly one instance of the played card from our hand
    let played_idx = seat.hand.iter().position(|c| match (candidate, c) {
        (
            Candidate::PlaySuit { shape, value },
            Card::Suit {
                shape: cs,
                value: cv,
            },
        ) => cs == shape && cv == value,
        (Candidate::PlayWhot { .. }, Card::Whot) => true,
        _ => false,
    });
    let Some(played_idx) = played_idx else {
        return 0.0;
    };

    let remaining: Vec<Card> = seat
        .hand
        .iter()
        .enumerate()
        .filter(|(i, _)| *i != played_idx)
        .map(|(_, &c)| c)
        .collect();

    // Playing last card — always right
    if remaining.is_empty() {
        return 50.0;
    }

    let follow_ups = valid_moves(&remaining, simulated_top, &None, ctx.state.mode);
    let scale = 1.0 / (ctx.state.seats.len() as f64 - 1.0).max(1.0);

    if follow_ups.is_empty() {
        if let Candidate::PlaySuit { value, .. } = candidate {
            if is_action_card(*value) {
                return 0.0;
            }
        }
        return -20.0 * scale;
    }

    (follow_ups.len() as f64 * 1.5).min(6.0) * scale
}
